### DISCUSSION TITLE
Implementing uScript for secure client-side scripting

### DISCUSSION THREAD

**Discussion by DanielWillett** [SmartlyDressedGames/Unturned-3.x-Community#3930] [Ideas]
The issue (#3921) opened the other day gave me the idea of reworking uScript as a client-side scripting language for servers to use. Obviously, it'd depend on the dev of uScript (Ster)'s feelings about letting you implement it into the game, I've got no connection to uScript and haven't discussed this with anyone on the dev team, but if they're fine with it then this could be a somewhat easy solution to client-side scripting.

If you don't know, uScript is a scripting language written for Unturned similar to JavaScript and C# and is compiled on load (not interpreted) so performance is not really a problem. A lot of server owners are already familiar with the syntax.

This could also make a UI rework not really needed as much.

Distribution would probably just be in workshop mods, maybe give them their own asset which can be loaded by the server then unloaded on disconnect, and to reference the script in RPC calls to and from the scripts.

### Why we need this
Server plugins can do a lot, but sometimes there's just one small thing you can't do, because there's just one variable you don't have access to or a little method call that's needed to update something, etc, and sure we could request events or methods on here for everything but realistically a server-side API can't cover all bases without just becoming unmanageable and inefficient. The possibilities with client-side scripts are basically endless. A UI rework would barely be needed anymore.

Just a few use cases I can think of off the top of my head; attaching an effect to an object (smooth animation compared to what can be done on the server), smoother UI animations, advanced UI interactions, localized time display (UTC to local), other language support for server messages/ui (to system language).

### Concerns
* Security: If everything is done right and there aren't any methods exposed that shouldn't be, it shouldn't be a problem, maybe an infinite loop that crashes your game could be abused, but it's not really a big deal, just don't join the server.
* API Documentation: Sure it would be nice but I wouldn't say its a necessity, at least at the initial release. I mean almost none of the unturned C# code is documented so it would be okay to delay heavy documentation at the start.
* Cheating: If scripts are validated against the server's copies with hash verification like barricades already are, this won't be an issue. If you're already modifying the game to remove that validation there's no point in using a script to cheat.

### Some features that would be very nice to have
* Something on the server homepage to show that scripts are enabled.
* Hash validation of scripts with the server (to avoid easy cheating without tripping battleye).
* Invoking methods (with arguments) from the server.
* The ability to send data back to the server.
* Direct manipulation of unity game objects for barricades, structures, vehicles, dropped items, players, objects.
  * Finding/setting children, parents, etc.
  * Starting animations, setting position, rotation, scale, etc (and the RectTransform equivalents)
  * With UI, directly modifying UI components, progress bars, etc with smoother animation.
  * Perhaps a way to create coroutines (?) idk how possible this is.
* Numerous methods and events to interact with UI, as this will be the main usage, setting text, check box states, responding to UI events (maybe a way to just subscribe to UnityEvents).
* Setting images from asset bundles or item icons.
* Expose a lot of the Interactable API (requests, etc).
* Expose a lot of LevelObejct methods (right now there's almost no manipulation from the server).
* Expose instantiating objects from master bundles.
* Remove all the web and MySQL stuff.
* Remove RocketMod reliance.
* Full uScript types for all assets with all asset fields.

Here's an invite to their discord if you do want to check it out. https://discord.gg/KaaEP9tUGd

**Comment by GazziFX** [comment]
Also hot reload for devs

**Comment by SDGNelson** [comment]
It would be awesome to have scripting support, but I don't see it happening anytime soon unless I either make Unturned's source available (maybe) or bring on another developer (maybe). For me it's currently some relief that I can make breaking changes to the client API, whereas server API changes slow everything down. When I rewrite the inventory it will have to maintain support for all the existing dumb inventory functionality for the sake of server plugins.

**Reply by fahrettinenes** [reply]
What's wrong with making Unturned open-source?

**Reply by DanielWillett** [reply]
Have you considered just releasing one large, plugin-breaking update? And just putting it on the preview branch for a month or two to give devs time? If it's on the preview branch for long enough people will have time to migrate and new plugins will be made to replace the old, unmaintained ones. If it's to the point where you're spending so much time catering to plugins written 4 years ago maybe it would be worth it to just remove all the obsolete code and rewrite some stuff that you want to rewrite without having to worry about backward compatibility for one update (state arrays, inventory, etc., no clue how you would make a state array rework backward compatable).

**Reply by liebesleid-eth** [reply]
I would love the idea of announcing a large game-breaking update and giving server owners - plugin developers time to swith and move out of all these 8 year old plugins

**Reply by Corbyncc** [reply]
> Have you considered just releasing one large, plugin-breaking update? And just putting it on the preview branch for a month or two to give devs time? If it's on the preview branch for long enough people will have time to migrate and new plugins will be made to replace the old, unmaintained ones. If it's to the point where you're spending so much time catering to plugins written 4 years ago maybe it would be worth it to just remove all the obsolete code and rewrite some stuff that you want to rewrite without having to worry about backward compatibility for one update (state arrays, inventory, etc., no clue how you would make a state array rework backward compatable).

Imma keep it a stack, hes right - id rather see breaking changes that require plugin devs to update their plugins than you waste your time trying to keep backwards compatibility for legacy code and bad plugins that server owners shouldn't be using in the first place.

> or bring on another developer (maybe)

Where should I send my resume? ;)

**Reply by SDGNelson** [reply]
I suppose it's true there could be an intentionally breaking update. I think a lot of players and server owners would be upset. Maybe the upset would be worth it? 🤔

**Reply by liebesleid-eth** [reply]
We complain anyways on each update 😭

**Reply by DanielWillett** [reply]
If it means big updates come quicker I personally wouldn't mind rewriting a few more lines of code than normal every 3 or 4 months.

I think something that could help with this is a bit more communication about big updates on the preview branch, and if you can make it part of a large rewrite, taking some of the third-party code out of the main assembly so you can publish the source code. (I may be wrong but I remember reading somewhere that's why you haven't open-sourced it yet). I remember the net-code rewrite PSA you made a while back, if we had something like that for all breaking API changes, checking the issue list every once in a while isn't that hard.

Otherwise, I can't see much logic behind getting seriously upset about a breaking update (not saying it won't happen, it definitely will). You have (as far as I know) never given any guarantee to not break stuff with the API. And for most complex plugins they have to be recompiled at least every few months from an update anyway because of an out-of-date patch or reflection. If it makes it easier for you to make new features I say go for it. As Corbyn said servers still using old plugins that barely work just need to move on. You spending so much extra time working around old systems just isn't worth it.

**Reply by Corbyncc** [reply]
> I suppose it's true there could be an intentionally breaking update. I think a lot of players and server owners would be upset. Maybe the upset would be worth it? 🤔

Breaking update for large legacy code rewrite is very welcome and would be much appreciated. As someone mentioned earlier, we complain every update anyways lmao

**Reply by Jdance-Media** [reply]
Definitely agree with Dan's (BlazingFlame) idea. Also promotes the creation of newer plugins that follow standards that were established more recently.

**Reply by Paradox304** [reply]
Don't mind this. I have written a large amount of plugins that mess with inventory and such. And wouldnt mind a large game breaking update just to fix the old inventory code, introducing instance ids to items, and the state array rework. As long as ample time to prepare is given and it is properly announced everywhere. In-game, discord, steam, etc.
